### ISSUE TITLE
Using `@contextlib.asynccontextmanager` to decorate an asynchronous generator function should be a valid use case

### ISSUE THREAD

**Issue by MapleCCC** [microsoft/pyright#2214]
**Describe the bug**
Using `@contextlib.asynccontextmanager` to decorate an asynchronous generator function should be a valid use case

**To Reproduce**
[An example from the standard library document](https://docs.python.org/3/library/contextlib.html#contextlib.asynccontextmanager) of `@contextlib.asynccontextmanager`:

```python
from collections.abc import Iterator
from contextlib import asynccontextmanager

class Connection: ...

async def acquire_db_connection() -> Connection: ...
async def release_db_connection(conn: Connection) -> None: ...

@asynccontextmanager
async def get_connection() -> Iterator[Connection]:
    conn = await acquire_db_connection()
    try:
        yield conn
    finally:
        await release_db_connection(conn)
```

Pyright complains that `Type "() -> Coroutine[Any, Any, Iterator[Connection]]" cannot be assigned to type "Callable[_P@asynccontextmanager, AsyncIterator[_T@asynccontextmanager]]"`.

**Expected behavior**
There should be no error for the above code.

**VS Code extension or command-line**
Pyright command-line 1.1.163

**Comment by erictraut** [comment]
Pyright is correct to flag this as an error. The `@asynccontextmanager` decorator requires that the decorated method return an `AsyncIterator`, not an `Iterator`. Mypy emits the same error as pyright here.

If you change your code to this, it will type check fine in both pyright and mypy.

```python
@asynccontextmanager
async def get_connection() -> AsyncIterator[Connection]:
```
